keps/sig-network/0029-nodelocal-dns-cache.md

```diff
@@ -0,0 +1,145 @@
+
+* With the current DNS achitecture, it is possible that pods with the highest DNS QPS have to reach out to a different node, if there is no local kube-dns instance.  
+Having a local cache will help improve the latency in such scenarios. 
+
+* Skipping iptables DNAT and connection tracking will help reduce [conntrack races](https://github.com/kubernetes/kubernetes/issues/56903) and avoid UDP DNS entries filling up conntrack table.
+
+* Connections from local caching agent to kube-dns can be upgraded to TCP. TCP conntrack entries will be removed on connection close in contrast with UDP entries that have to timeout ([default](https://www.kernel.org/doc/Documentation/networking/nf_conntrack-sysctl.txt) `nf_conntrack_udp_timeout` is 30 seconds)
+
+* Metrics & visibility into dns requests at a node level.
+
+* Neg caching can be re-enabled, thereby reducing number of queries to kube-dns.
+
+* There are several open github issues proposing a local DNS Cache daemonset and scripts to run it:
+	* [https://github.com/kubernetes/kubernetes/issues/7470#issuecomment-248912603](https://github.com/kubernetes/kubernetes/issues/7470#issuecomment-248912603)
+
+	* [https://github.com/kubernetes/kubernetes/issues/32749](https://github.com/kubernetes/kubernetes/issues/32749)
+
+	* [https://github.com/kubernetes/kubernetes/issues/45363](https://github.com/kubernetes/kubernetes/issues/45363)
+
+
+This shows that there is interest in the wider Kubernetes community for a solution similar to the proposal here. 
+
+
+### Goals
+
+Being able to run a dns caching agent as a Daemonset and get pods to use the local instance. Having visibility into cache stats and other metrics.
+
+### Non-Goals
+
+Providing a replacement for kube-dns.
+
+## Proposal
+
+A nodeLocal dns cache runs on all cluster nodes. This is managed as an add-on, runs as a Daemonset. All pods using clusterDNS will now talk to the nodeLocal cache, which will query kube-dns in case of cache misses in cluster.local domain. The node's resolv.conf will be used by this local agent for all other cache misses. 
+
+#### Daemonset, Service, Listen Interface for caching agent
+
+The caching agent daemonset runs in hostNetwork mode in kube-system namespace with a Priority Class of “system-node-critical”. It listens for dns requests on a dummy interface created on the host. A separate ip address is assigned to this dummy interface, so that requests to kube-dns or any other custom service are not incorrectly intercepted by the caching agent. This ip address is obtained by creating a nodelocaldns service, with no endpoints. Kubelet takes the service name as an argument `--localDNS=<namespace>/<svc name>`, looks up the ip address and populates pods' resolv.conf with this value instead of clusterDNS. Each cluster node will have a dummy interface with this service ip assigned to it. The Daemonset runs in privileged securityContext since it needs to create this dummy interface and add iptables rules.
+
+#### iptables NOTRACK
```

Comment:
we need to think about how this would or would-not work with ipvs kube-proxy, and document that VERY CLEARLY.  If this only works with iptables mode, we probably want to think harder about it.

We should maybe get some guidance from non-core implementations of services - could they work in a mode like this?

I don't think we have to make this a universal solution - if it only works with the upstream kube-proxy, that is OK.  Probably.  But we need to be confident how widely this technique applies vs just the idea being translatable.

Reply:
I haven't tried this with ipvs mode, but since the exposed service has no endpoints, there shouldn't be a difference. I will try it out.

Reply:
@m1093782566

Reply:
@prameshj You can ping me if you have any trouble in trying IPVS mode.

Reply:
Thanks, I will reach out.